library/core/src/primitive_docs.rs

```diff
@@ -268,6 +268,52 @@ mod prim_bool {}
 /// [`Debug`]: fmt::Debug
 /// [`default()`]: Default::default
 ///
+/// # Never type fallback
+///
+/// When the compiler sees a value of type `!` it implicitly inserts a coercion (if possible)
+/// to allow type checker to infer any type:
+///
+/// ```rust,ignore (illustrative-and-has-placeholders)
+/// // this
+/// let x: u8 = panic!();
+///
+/// // is (essentially) turned by the compiler into
+/// let x: u8 = absurd(panic!());
+///
+/// // where absurd is a function with the following signature
+/// // (it's sound, because `!` always marks unreachable code):
+/// fn absurd<T>(_: !) -> T { ... }
+// FIXME: use `core::convert::absurd` here instead, once it's merged
+/// ```
+///
+/// While it's convenient to be able to use non-diverging code in one of the branches (like
+/// `if a { b } else { return }`) this could lead to compilation errors:
+///
+/// ```compile_fail
+/// // this
+/// { panic!() };
+///
+/// // gets turned into this
+/// { absurd(panic!()) }; // error: can't infer the type of `absurd`
+/// ```
+///
+/// To prevent such errors, the compiler remembers where it inserted `absurd` calls, and
+/// if it can't infer their type, it sets the type to the fallback type:
```

Comment:
```suggestion
/// if it can't infer their return type, it sets that type to the fallback type:
```

Reply:
"return type" feels like it could hypothetically be confusing because they might think "return type from this function" but it actually is "return type from the injected phantasmal `absurd(expr)`, which resolves to whatever is at the binding site, if possible".

no strong recommendation on a replacement though.

Reply:
I think `their type` is fine (or maybe `the type`), since that is how it is referenced in `sets the type to` and also in my suggested change above.

Reply:
I also found "return type" to be more confusing. I think "the type" is pretty good, but it does mean the sentence has a lot of "type"s in it. You could go with something like this if you find that problematic:
```suggestion
/// if it can't infer the type, it uses the fallback type instead:
```